### PR TITLE
Use PackedStruct to store optional timeouts in Route

### PR DESCRIPTION
**Pull request by yurykats**
In #24659 optional Route timeouts were optimized by using a struct instead of `absl::optional`. Since then, a more efficient `PackedStruct` was introduced, that provides even more efficiency, especially in the cases when some timeouts are set, but not all. The PackedStruct is  storing only set values, and not waisting 8 bytes on each of them being optional.

This PR is switching Routes from using a internal struct to using PackedStruct.

This is a step towards more memory efficient config data structures. Issue #24154.

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by pradeepcrao — COMMENTED** [review]
Thanks so much for doing this Yury!

File: source/common/router/config_impl.cc [hunk 1]
```diff
@@ -1146,13 +1109,36 @@ std::unique_ptr<InternalRedirectPolicyImpl> RouteEntryImplBase::buildInternalRed
   return std::make_unique<InternalRedirectPolicyImpl>(policy_config, validator, current_route_name);
 }
 
-std::unique_ptr<OptionalTimeouts> RouteEntryImplBase::buildOptionalTimeouts(
+RouteEntryImplBase::OptionalTimeouts RouteEntryImplBase::buildOptionalTimeouts(
     const envoy::config::route::v3::RouteAction& route) const {
-  if (route.has_idle_timeout() || route.has_max_grpc_timeout() || route.has_grpc_timeout_offset() ||
-      route.has_max_stream_duration()) {
-    return std::make_unique<OptionalTimeouts>(route);
+  OptionalTimeouts timeouts;
```
[the remark this answers is not in the document]
Comment (yurykats) [reply source/common/router/config_impl.cc:1114]: It is mentioned in the packed_struct.h, but I didn't think this was critical. Can certainly pre-calculate the number of optional values.

[the remark this answers is not in the document]
Comment (yurykats) [reply source/common/router/config_impl.cc]: Done

File: source/common/router/config_impl.h [hunk 2]
```diff
@@ -723,27 +670,65 @@ class RouteEntryImplBase : public RouteEntryAndRoute,
...
+    GrpcTimeoutHeaderMax,
+    GrpcTimeoutHeaderOffset,
+    MaxGrpcTimeout,
+    GrpcTimeoutOffset
+  };
+  using OptionalTimeouts = PackedStruct<std::chrono::milliseconds, 6, OptionalTimeoutNames>;
+
   absl::optional<std::chrono::milliseconds> idleTimeout() const override {
-    return optional_timeouts_ != nullptr ? optional_timeouts_->idleTimeout() : absl::nullopt;
+    auto timeout = optional_timeouts_.get<OptionalTimeoutNames::IdleTimeout>();
```
[the remark this answers is not in the document]
Comment (yurykats) [reply source/common/router/config_impl.h]: Done

**Comment by yurykats** [comment]
> LGTM, left a small nit. Also should think about @pradeepcrao's comment, and applying to the other PR that is using PackedStruct

Addressed @pradeepcrao's comment here, and will also apply to Cluster timeouts.

**Comment by yurykats** [comment]
PTAL

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

/assign-from @envoyproxy/senior-maintainers

File: source/common/router/config_impl.h [hunk 3]
```diff
@@ -750,26 +697,63 @@ class RouteEntryImplBase : public RouteEntryAndRoute,
...
+    GrpcTimeoutHeaderMax,
+    GrpcTimeoutHeaderOffset,
+    MaxGrpcTimeout,
+    GrpcTimeoutOffset
+  };
+  using OptionalTimeouts = PackedStruct<std::chrono::milliseconds, 6, OptionalTimeoutNames>;
+
   absl::optional<std::chrono::milliseconds> idleTimeout() const override {
-    return optional_timeouts_ != nullptr ? optional_timeouts_->idleTimeout() : absl::nullopt;
+    const auto timeout = optional_timeouts_.get<OptionalTimeoutNames::IdleTimeout>();
```
Comment (ggreenway) [inline source/common/router/config_impl.h:717]: Please remove duplication of these 6 accessors with a helper function (which may need to be a template function)

**Review by ggreenway — CHANGES_REQUESTED** [review]
Looks nice overall

/wait

File: source/common/router/config_impl.h (hunk 3 above)
Comment (yurykats) [reply source/common/router/config_impl.h]: Done

**Review by ggreenway — APPROVED** [review]
LGTM

**Review by adisuissa — APPROVED** [review]
LGTM, thanks!

**Pull request merged**